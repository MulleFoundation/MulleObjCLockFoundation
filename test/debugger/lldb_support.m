#import <MulleObjC/MulleObjC.h>
#import <mulle-objc-runtime/mulle-objc-lldb.h>

#ifndef DEBUG
#define DEBUG 1
#endif


@interface Foo : NSObject

- (void) a;
+ (void) a;

@end


@interface Bar : Foo
@end



@implementation Foo

+ (void) a
{
   printf( "%s\n", __PRETTY_FUNCTION__);
}


- (void) a
{
   printf( "%s\n", __PRETTY_FUNCTION__);
}


@end


@implementation Bar

+ (void) a
{
   printf( "%s\n", __PRETTY_FUNCTION__);
}


- (void) a
{
   printf( "%s\n", __PRETTY_FUNCTION__);
}

@end


void  test_obj_class( id obj, Class cls)
{
   IMP    imp;

   imp    = mulle_objc_lldb_lookup_implementation( obj, @selector( a), cls, 0, 0, DEBUG);
   (*imp)( obj, @selector( a), NULL);
}



void  test_class_class( id obj, Class cls)
{
   IMP    imp;

   imp    = mulle_objc_lldb_lookup_implementation( obj, @selector( a), cls, 0, 1, DEBUG);
   (*imp)( obj, @selector( a), NULL);
}



void  test_obj_classid( id obj, mulle_objc_classid_t clsid)
{
   IMP    imp;

   imp    = mulle_objc_lldb_lookup_implementation( obj, @selector( a), (void *) clsid, 1, 0, DEBUG);
   (*imp)( obj, @selector( a), NULL);
}


void  test_class_classid( id obj, mulle_objc_classid_t clsid)
{
   IMP    imp;

   imp    = mulle_objc_lldb_lookup_implementation( obj, @selector( a), (void *) clsid, 1, 1, DEBUG);
   (*imp)( obj, @selector( a), NULL);
}



main()
{
   Bar    *bar;
   Class  barCls;
   Class  fooCls;

   barCls = [Bar class];
   fooCls = [Foo class];
   bar    = [Bar new];

   test_obj_class( bar, _mulle_objc_object_get_isa( bar));
   test_class_class( barCls, _mulle_objc_object_get_isa( barCls));

   test_obj_class( bar, _mulle_objc_class_get_superclass( _mulle_objc_object_get_isa( bar)));
   test_class_class( barCls, _mulle_objc_class_get_superclass( _mulle_objc_object_get_isa( barCls)));

   test_obj_classid( bar,    @selector( Bar));
   test_class_classid( barCls, @selector( Bar));

   test_obj_classid( bar,    @selector( Foo));
   test_class_classid( barCls, @selector( Foo));

   [bar release];
}
