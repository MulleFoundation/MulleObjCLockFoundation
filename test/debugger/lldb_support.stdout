-[Bar a]
+[Bar a]
-[Foo a]
+[Foo a]
-[Bar a]
+[Bar a]
-[Foo a]
+[Foo a]
